rfcs/20200810-tf-internal-api.md

```diff
@@ -0,0 +1,119 @@
+   of TensorFlow, or to integrate such a framework into TensorFlow's API.
+1. Mature enough and won't change behavior/interface for every release. 
+1. Non-trivial, otherwise the function should be copied to the client side.
+1. Widely used in the implementation of TF public APIs (i.e. new functionality 
+   isn’t immediately added to the tf.internal namespace)
+1. Has at least two downstream libraries which are known to need it.
+
+TF API owners will review the new API proposal, and follow the existing review 
+process for core TF API.
+
+
+### Documentation and Docstring
+The "internal" API should have the same style and standard as the core 
+TensorFlow API, which is documented [here](https://github.com/tensorflow/community/blob/master/governance/api-reviews.md#docstrings). 
+We should explicitly list out the difference between "internal" API and core
+API, and also choose a different place on tensorflow.org so that existing user 
+are not confused.
+
+### Naming and sub namespace
+Similar "internal" APIs should be grouped together as sub namespaces, e.g., test 
+related APIs should live under "tf.internal.test". This is aligned with the 
+existing TF naming conversion.
+
+Try not to export experimental APIs since the "internal" API should be mature 
+enough.
+
+### Current candidate
+The following list is created from the private TF method usage within Keras, 
+when we were trying to convert Keras to use the public TF API only. This is by 
+no means the full list, but will serve as the first wave of review requests we 
+send to the API owner for review. We don't expect all of them to be approved, 
+and will discuss with the API owner on a case to case basis.
+
+|Symbol location  |API Name  | 
+:-------------- |:---------------------------------------------------- |
+|python.framework.func_graph.FuncGraph |tf.internal.FuncGraph     |
+|python.framework.combinations.*| tf.internal.test.combinations.* |
+|python.distribute.strategy_combinations.* |tf.internal.test.combinations.* |
+|python.training.tracking.base.no_automatic_dependency_tracking|tf.internal.tracking.no_automatic_dependency_tracking |
+|python.util.object_identity.*|tf.internal.object_identity.* |
+|python.util.tf_decorator.*|tf.internal.decorator.* |
+|python.util.tf_inspect.*|tf.internal.inspect.* |
+
+### Alternative Names
+1. <b>"tf.internal"</b>: It gives the user the impression that this is not a 
+   standard user facing API (good), and contains some private/implementation 
+   details that are internal to TF. 
+1. <b>"tf.infra"</b>: infrastructure is aligned with "building blocks" and low 
+   level functionalities, like file system/network etc. So far, the APIs we want
+   to add are still high level APIs and utility functions. 
+1. By Martin <b>"tf._internal"</b>: the extra "_" emphasis in the pythonic way
+   that this is for private usage.
+
+
+## Questions and Discussion Topics
+
+1. Naming of the API namespace.
```

Comment:
I want to suggest we do not expose these symbols as a public namespace, and instead use "_internal". This would avoid creating another carveout from semver, and the effect for downstream projects is the same (they have to know what they're doing). 

Alternatively, we could expose these as a module separate from tensorflow (similar to tfdbg or other projects are in separate modules).

Reply:
I think the naming of the namespace will be the biggest topic for the design meeting, and the "tf._intenral" has been listed as one of the option here.

Since those module would have dependency from and to TF code, its code won't be able to split out from TF repo. I am not sure if they could live under a different named PIP package (if this is what you mean for a "separate module").